core/sequencer/runner/native.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package runner contains methods for running a pipeline.
+package runner
+
+import (
+	"github.com/golang/glog"
+
+	"github.com/google/keytransparency/core/mutator"
+	"github.com/google/keytransparency/core/sequencer/mapper"
+
+	pb "github.com/google/keytransparency/core/api/v1/keytransparency_go_proto"
+	tpb "github.com/google/trillian"
+)
+
+// ApplyMutations takes the set of mutations and applies them to given leafs.
+// Returns a list of map leaves that should be updated.
+func ApplyMutations(mutatorFunc mutator.ReduceMutationFn,
+	msgs []*pb.EntryUpdate, leaves []*tpb.MapLeaf) ([]*tpb.MapLeaf, error) {
+	// Index the updates.
+	indexedUpdates, err := DoMapUpdateFn(mapper.MapUpdateFn, msgs)
+	if err != nil {
+		return nil, err
+	}
+
+	joined := Join(leaves, indexedUpdates)
+
+	ret := make([]*tpb.MapLeaf, 0, len(joined))
+	for _, j := range joined {
+		mapper.ReduceFn(mutatorFunc, j.Index, j.Leaves, j.Msgs,
+			func(l *tpb.MapLeaf) { ret = append(ret, l) })
+	}
+	glog.V(2).Infof("ApplyMutations applied %v mutations to %v leaves", len(msgs), len(leaves))
+	return ret, nil
+}
+
+// Joined is the result of a CoGroupByKey on []*MapLeaf and []*IndexUpdate.
+type Joined struct {
+	Index  []byte
+	Leaves []*tpb.MapLeaf
```

Comment:
Should it not be possible to have multiple leaves with the same index?

Reply:
Depends on which way we look at it. 
From the point of view of the map which we know should only produce one leaf per index, yes. 
From the perspective of a map reduce, the mapping step can emit multiple <K,V> pairs for any given K.